cachi2/core/package_managers/pip.py

```diff
@@ -1403,71 +1420,92 @@ def _download_dependencies(
+            if dpi.should_verify_checksums():
+                _hash_verify(dpi.path, dpi.checksums_to_verify)
 
-            download_binary_file(source.url, source.path, auth=None)
-            _check_metadata_in_sdist(source.path)
-            download_info = source.download_info
+            if dpi.package_type == "sdist":
+                _check_metadata_in_sdist(dpi.path)
 
-        elif req.kind == "vcs":
-            download_info = _download_vcs_package(req, pip_deps_dir)
-        elif req.kind == "url":
-            download_info = _download_url_package(req, pip_deps_dir, trusted_hosts)
         else:
-            # Should not happen
-            raise RuntimeError(f"Unexpected requirement kind: {req.kind!r}")
+            if require_hashes or req.kind == "url":
+                hashes = req.hashes or [req.qualifiers.get("cachito_hash", "")]
+                _hash_verify(download_info["path"], list(map(_to_checksum_info, hashes)))
+
+        download_info["hash_verified"] = True
 
         log.info(
-            "Successfully downloaded %s to %s",
+            "Successfully downloaded/checked '%s' in path '%s'",
             req.download_line,
             download_info["path"].relative_to(output_dir),
         )
 
-        if require_hashes or req.kind == "url":
-            hashes = req.hashes or [req.qualifiers["cachito_hash"]]
-            must_match_any_checksum(download_info["path"], list(map(_to_checksum_info, hashes)))
-            download_info["hash_verified"] = True
-        else:
-            download_info["hash_verified"] = False
+        return download_info
 
-        download_info["kind"] = req.kind
-        download_info["requirement_file"] = str(requirements_file.file_path.subpath_from_root)
-        downloaded.append(download_info)
+    for req in requirements_file.requirements:
+        log.info("Processing requirement line '%s'", req.download_line)
 
-    if allow_binary:
-        log.info("Downloading %d wheel(s) ...", len(to_download))
-        files: dict[str, Union[str, PathLike[str]]] = {pkg.url: pkg.path for pkg in to_download}
-        asyncio.run(async_download_files(files, get_config().concurrency_limit))
+        if req.kind == "pypi":
+            artifacts = _process_package_distributions(req, pip_deps_dir, allow_binary)
+            artifacts_to_download, artifacts_to_check = _classify_download_or_local(artifacts)
 
-        for pkg in to_download:
-            try:
-                if pkg.should_verify_checksums():
-                    must_match_any_checksum(pkg.path, pkg.checksums_to_verify)
-            except PackageRejected:
-                pkg.path.unlink()
-                log.warning("The %s is removed from the output directory", pkg.path.name)
+            for dpi in artifacts_to_download:
+                log.info("Processing pypi download '%s'", dpi.path.name)
+                asyncio.run(
+                    async_download_files({dpi.url: dpi.path}, get_config().concurrency_limit)
```

Comment:
Are we now downloading all of the artifacts one-by-one? Previously we were processing a list of wheels asynchronously

Reply:
I need to take another look at this.